src/main/java/com/microsoft/azure/spring/data/cosmosdb/core/DocumentDbTemplate.java

```diff
@@ -443,21 +453,51 @@ private RequestOptions getRequestOptions(PartitionKey key, Integer requestUnit)
-        Assert.notNull(domainClass, "domainClass should not be null.");
-        Assert.hasText(collectionName, "collection should not be null, empty or only whitespaces");
+    @Override
+    public <T> List<T> find(@NonNull DocumentQuery query, @NonNull String collectionName,
+                            @NonNull Class<T> entityClass, @NonNull List<String> partitionKeyNames) {
+        validate(query, collectionName, entityClass, partitionKeyNames);
+        validateQuery(query, entityClass, collectionName);
 
-        validateQuery(query, domainClass, collectionName);
+        return Lists.newArrayList(
+                findAsync(query, collectionName, entityClass, partitionKeyNames).toBlocking().getIterator());
+    }
 
-        final com.microsoft.azure.documentdb.SqlQuerySpec sqlQuerySpec = new FindQuerySpecGenerator().generate(query);
-        final boolean isCrossPartitionQuery = query.isCrossPartitionQuery(getPartitionKeyNames(domainClass));
+    @Override
+    public <T> Observable<T> findAsync(@NonNull DocumentQuery query, @NonNull String collectionName,
+                                       @NonNull Class<T> entityClass, List<String> partitionKeyNames) {
+        validate(query, collectionName, entityClass, partitionKeyNames);
 
-        return this.executeQuery(sqlQuerySpec, isCrossPartitionQuery, domainClass, collectionName);
+        final FeedOptions feedOptions = new FeedOptions();
+        final SqlQuerySpec sqlQuerySpec = new FindQuerySpecGenerator().generateAsync(query);
+
+        feedOptions.setEnableCrossPartitionQuery(query.isCrossPartitionQuery(partitionKeyNames));
+
+        return executeQueryAsync(sqlQuerySpec, collectionName, feedOptions)
+                .map(FeedResponse::getResults)
+                .flatMap(Observable::from)
+                .map(d -> this.getConverter().readAsync(entityClass, d));
     }
 
-    public <T> Boolean exists(@NonNull DocumentQuery query, @NonNull Class<T> domainClass, String collectionName) {
-        return this.find(query, domainClass, collectionName).size() > 0;
+    @Override
+    public <T> Boolean exists(@NonNull DocumentQuery query, @NonNull String collectionName,
+                              @NonNull Class<T> entityClass, @NonNull List<String> partitionKeyNames) {
+        validate(query, collectionName, entityClass, partitionKeyNames);
+
+        return existsAsync(query, collectionName, entityClass, partitionKeyNames)
+                .toBlocking()
+                .single();
+    }
+
+    @Override
+    public <T> Observable<Boolean> existsAsync(@NonNull DocumentQuery query, @NonNull String collectionName,
+                                               @NonNull Class<T> entityClass, @NonNull List<String> partitionKeyNames) {
+        validate(query, collectionName, entityClass, partitionKeyNames);
+
+        return findAsync(query, collectionName, entityClass, partitionKeyNames)
+                .count()
+                .map(c -> c > 0)
+                .single();
```

Comment:
Can we use the `countAsync` method to check exists? `Find` might be more heavier than count only, not confirmed though.